doc/design/service-discovery.md

```diff
@@ -0,0 +1,228 @@
+# Service Discovery
+
+## Summary and problem description
+
+Service discovery over IPC channel e.g. message queue or UNIX domain socket is not performant since larger data is
+transferred, which can lead to transmission of several frames. If lots of services are discovered at high-frequency
+e.g. at startup the IPC channel can become a bottleneck. Furthermore, a lession learned from the iceoryx development
+so far is, that the IPC channel should just be used for the earliest communication during startup and not for the
+creation of objects inside the shared memory (see [this issue](https://github.com/eclipse-iceoryx/iceoryx/issues/611)).
+
+### Status quo in iceoryx Almond
+
+#### Runtime
+
+* `cxx::expected<InstanceContainer, FindServiceError> PoshRuntime::findService(capro::ServiceDescription)`
+  * Sends message over IPC channel
+
+#### RouDi
+
+* `runtime::IpcMessage {ProcessManager,PortManager}::findService(capro::ServiceDescription)`
+  * Called after RouDi has received a request from the runtime
+  * Sends back answer to `PoshRuntime` over IPC channel
+
+## Terminology
+
+## Requirements
+
+* Polling and event notification approach shall both be possible
+
+### Considerations and use-cases
+
+#### AUTOSAR Adaptive
+
+* findService() call as implemented in Almond v1.0
+* ara::com service discovery callbacks
+
+1. `StartFindService()`: Register callback, std::function interface
+1. `StopFindService()`: Unregister callback
+1. `FindService()`: one-shot, polling, synchronous
+
+#### SOME/IP-SD
+
+* Very similar to AUTOSAR Adaptive
+
+#### Cyclone DDS and ROS 2
+
+* Cyclone DDS calls this feature [topic discovery](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/topic_discovery.md). See also the [feature request](https://github.com/eclipse-cyclonedds/cyclonedds/issues/42).
+
+* Based on SDEP (simple endpoint discovery protocol)
+* `dds_ktopic`
+  * stores pointer to `ddsi_topic_definition`
+* Has built-in DCPSTopic topic and contains
+  * Topic name
+  * Type name
+  * Topic key
+  * QoS settings
+* `dds_find_topic_locally` function is available for users
+
+#### Static discovery
+
+* Static discovery shall be supported
+  * Having RouDi read in a config file, which has
+    * All the topics listed
+    * Wiring pub/sub information
```

Comment:
> Does RouDi create all the publishers/subscribers during initialization and then just provides the fitting subscriber/publisher to the application?

Without having thought too much about the details: Yep, what you describe is what I have in my head regarding the static shared memory feature.

> Or are the services just registered in the service registry and not really present? If so what happens when the subscriber assumes that a publisher is present but it is not and therefore cannot subscribe to it?

What you describe here is the scenario where an application is not yet running but the static discovery returns information about pub/subs of this runtime, right? Instead of creating a new interface / API here, I would opt for a `{stop}offer` and `{un}subscribe` that are no-ops in the static case. Same [reason](https://github.com/eclipse-iceoryx/iceoryx/pull/784#discussion_r633514476) as before. I would say it's not misuse, but instead a feature, which allows the user a smooth transition from the dynamic world to a static world.

Reply:
Roudi should still react as before on Offer/StopOffer requests I think. StopOffer should not really be used by applications in the static case (except for the publisher dtor in the shutdown phase of the aplication). If you do so, the application is not designed for static discovery. Changing the behavior to no op does not change much, the application also just stop publishing data (you cannot force it to send something, even if it looks/is connected).

Unsubscribe is not really an issue, if the application does not want the data it is free to unsubscribe. The point of  static discovery is in my opinion that the system is designed in a way that the available services do not change (after a startup phase until a shutdown phase). Whether you subscribe to them or not is a separate question.

The main problem is if an app does not follow this contract and stops offering a previously offered services mid-run. Then the static discovery information does not get updated and users trying to subscribe will just not find the service (and hence receive no data). In the static discovery case Roudi or some monitor must ensure that we react to this scenario (e.g. print a warning, terminate, something in-between).

Reply:
> StopOffer should not really be used by applications in the static case (except for the publisher dtor in the shutdown phase of the aplication). If you do so, the application is not designed for static discovery. Changing the behavior to no op does not change much, the application could also just stop publishing data (you cannot force it to send something, even if it looks/is connected).

Well, it's complicated. We need to provide users with a smooth transition from a dynamic world into a static world. So I agree, `stopOffer` should not be used but will likely for simplicity just become a no-op.

I'll added a link to this discussion in #539, we can continue there.